### PR TITLE
fix(float): missing default highlight for title

### PR DESCRIPTION
**Pull request by glepnir**
Problem: there is missing default title highlight when highlight not defined in title text chunk.

Solution: when attr is not set use default title highlight group.

### CODE REVIEW

**Review by zeertzjq — DISMISSED** [review]
This is not documented.

**Comment by zeertzjq** [comment]
I'm also not sure if this is desired. This is not backward-compatible with v0.9.

**Comment by glepnir** [comment]
When no highlight is configured in  chunk , assume they use the default title highlight like string.

File: runtime/doc/api.txt [hunk 1]
```diff
@@ -3270,13 +3270,15 @@ nvim_open_win({buffer}, {enter}, {config})                   *nvim_open_win()*
 <
                   • title: Title (optional) in window border, string or list.
                     List should consist of `[text, highlight]` tuples. If
-                    string, the default highlight group is `FloatTitle`.
+                    string or not config highlight in tuples, the default
```
Comment (justinmk) [inline runtime/doc/api.txt:3273]:
```suggestion
                    string, or the tuple lacks a highlight, the default
```

File: src/nvim/drawscreen.c [hunk 2]
```diff
@@ -715,14 +716,17 @@ void end_search_hl(void)
...
-static void win_redr_bordertext(win_T *wp, VirtText vt, int col)
+static void win_redr_bordertext(win_T *wp, VirtText vt, int col, BorderTextType bt)
 {
   for (size_t i = 0; i < kv_size(vt);) {
     int attr = 0;
     char *text = next_virt_text_chunk(vt, &i, &attr);
     if (text == NULL) {
       break;
     }
+    if (attr == 0) {
```
Comment (bfredl) [inline src/nvim/drawscreen.c:727]: if attr was _explicitly_ set to zero by including a zero in the tuple like `[["item", 0], ...]` this must be respected.

File: src/nvim/decoration.c [hunk 3]
```diff
@@ -346,7 +346,14 @@ char *next_virt_text_chunk(VirtText vt, size_t *pos, int *attr)
   for (; text == NULL && *pos < kv_size(vt); (*pos)++) {
     text = kv_A(vt, *pos).text;
     int hl_id = kv_A(vt, *pos).hl_id;
-    *attr = hl_combine_attr(*attr, hl_id > 0 ? syn_id2attr(hl_id) : 0);
+    if (attr != NULL) {
```
Comment (zeertzjq) [inline src/nvim/decoration.c:349]: Why check for `NULL`?

File: src/nvim/decoration.c [hunk 4]
```diff
@@ -346,7 +346,14 @@ char *next_virt_text_chunk(VirtText vt, size_t *pos, int *attr)
...
+      *attr = hl_combine_attr(*attr, hl_id > 0 ? syn_id2attr(hl_id) : 0);
+      if (*attr == 0) {
+        // If no highlight is detected, set attr to -1 to indicate missing highlight
+        // for subsequent highlight detection.
+        *attr = -1;
+      }
```
Comment (zeertzjq) [inline src/nvim/decoration.c:355]: I don't think this is correct. How does this prevent overriding when `hl_id` is 0?

File: src/nvim/drawscreen.c [hunk 5]
```diff
@@ -715,14 +716,17 @@ void end_search_hl(void)
   screen_search_hl.rm.regprog = NULL;
 }
 
-static void win_redr_bordertext(win_T *wp, VirtText vt, int col)
+static void win_redr_bordertext(win_T *wp, VirtText vt, int col, BorderTextType bt)
 {
   for (size_t i = 0; i < kv_size(vt);) {
     int attr = 0;
```
Comment (zeertzjq) [inline src/nvim/drawscreen.c:722]:
```suggestion
    int attr = -1;
```

File: src/nvim/drawscreen.c [hunk 6]
```diff
@@ -715,14 +716,17 @@ void end_search_hl(void)
...
     char *text = next_virt_text_chunk(vt, &i, &attr);
     if (text == NULL) {
       break;
     }
+    if (attr == -1) {  // mean highlight is missing.
```
Comment (zeertzjq) [inline src/nvim/drawscreen.c:727]:
```suggestion
    if (attr == -1) {  // No highlight specified.
```

File: src/nvim/api/win_config.c [hunk 7]
```diff
@@ -189,13 +189,13 @@
 ///     ```
 ///   - title: Title (optional) in window border, string or list.
 ///     List should consist of `[text, highlight]` tuples.
-///     If string, the default highlight group is `FloatTitle`.
+///     If string, or the tuple lacks a highlight, the default highlight group is `FloatTitle`.
```
Comment (zeertzjq) [inline src/nvim/api/win_config.c:192]:
```suggestion
///     If string, or a tuple lacks a highlight, the default highlight group is `FloatTitle`.
```

File: src/nvim/api/win_config.c [hunk 8]
```diff
@@ -189,13 +189,13 @@
...
 ///   - title_pos: Title position. Must be set with `title` option.
 ///     Value can be one of "left", "center", or "right".
 ///     Default is `"left"`.
 ///   - footer: Footer (optional) in window border, string or list.
 ///     List should consist of `[text, highlight]` tuples.
-///     If string, the default highlight group is `FloatFooter`.
+///     If string, or the tuple lacks a highlight, the default highlight group is `FloatFooter`.
```
Comment (zeertzjq) [inline src/nvim/api/win_config.c:198]:
```suggestion
///     If string, or a tuple lacks a highlight, the default highlight group is `FloatFooter`.
```

File: runtime/doc/api.txt [hunk 9]
```diff
@@ -3270,13 +3270,15 @@ nvim_open_win({buffer}, {enter}, {config})                   *nvim_open_win()*
 <
                   • title: Title (optional) in window border, string or list.
                     List should consist of `[text, highlight]` tuples. If
-                    string, the default highlight group is `FloatTitle`.
+                    string, or the tuple lacks a highlight, the default
```
Comment (zeertzjq) [inline runtime/doc/api.txt:3273]:
```suggestion
                    string, or a tuple lacks a highlight, the default
```

File: runtime/doc/api.txt [hunk 10]
```diff
@@ -3270,13 +3270,15 @@ nvim_open_win({buffer}, {enter}, {config})                   *nvim_open_win()*
...
+                    highlight group is `FloatTitle`.
                   • title_pos: Title position. Must be set with `title`
                     option. Value can be one of "left", "center", or "right".
                     Default is `"left"`.
                   • footer: Footer (optional) in window border, string or
                     list. List should consist of `[text, highlight]` tuples.
-                    If string, the default highlight group is `FloatFooter`.
+                    If string, or the tuple lacks a highlight, the default
```
Comment (zeertzjq) [inline runtime/doc/api.txt:3280]:
```suggestion
                    If string, or a tuple lacks a highlight, the default
```

File: src/nvim/drawscreen.c [hunk 11]
```diff
@@ -715,14 +716,17 @@ void end_search_hl(void)
...
 {
   for (size_t i = 0; i < kv_size(vt);) {
-    int attr = 0;
+    int attr = -1;
     char *text = next_virt_text_chunk(vt, &i, &attr);
     if (text == NULL) {
       break;
     }
+    if (attr == -1) {  // No highlight specified.
+      attr = bt == kBorderTextTitle ? wp->w_ns_hl_attr[HLF_BTITLE] : wp->w_ns_hl_attr[HLF_BFOOTER];
```
Comment (bfredl) [inline src/nvim/drawscreen.c:728]: can reduce the scope of the ternary to `wp->w_ns_hl_attr[bt == kBorderTextTitle ? HLF_BTITLE : HLF_BFOOTER]

**Comment by bfredl** [comment]
Togheter with this change, we could get rid of `default_hl_id` in `parse_bordertext()` and use -1 there, right? Would be a deduplication of logic.

**Pull request merged**